open-api/lang-tools/src/main/java/io/deephaven/lang/completion/DelegatingCompletionHandler.java

```diff
@@ -31,37 +34,29 @@ public DelegatingCompletionHandler(CompletionHandler ... handlers) {
     }
 
     @Override
-    public CompletableFuture<? extends Collection<CompletionFragment>> complete(String command, int offset) {
-        // this is ...kind of silly... we should just be pushing results as they come in.
-        // TODO: a new api that supports push, with the future-based api adding blocking around said push messages.  IDS-1517-25
-        CompletableFuture[] futures = new CompletableFuture[handlers.length];
-        Collection<CompletionFragment>[] results = new Collection[handlers.length];
+    public Collection<CompletionItem.Builder> runCompletion(final ParsedDocument doc, final Position pos, final int offset) {
+        CompletableFuture<Collection<CompletionItem.Builder>>[] futures = new CompletableFuture[handlers.length];
         for (int i = 0; i < handlers.length; i++) {
-            final int slot = i;
             final CompletionHandler handler = handlers[i];
-            final CompletableFuture<? extends Collection<CompletionFragment>> started = handler.complete(command, offset);
-            futures[i] = CompletableFuture.supplyAsync(()->{
-                try {
-                    return results[slot] = started.get(5, TimeUnit.SECONDS);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt(); // stay interrupted
-                } catch (ExecutionException e) {
-                    // log this
-                    throw new RuntimeException("Failure in completion handler code for " + handler, e.getCause());
-                } catch (TimeoutException e) {
-                    // log this
-                    throw new RuntimeException("Timeout for completion handler code for " + handler, e.getCause());
-                }
-                return results[slot] = Collections.emptyList();
-            });
+            futures[i] = CompletableFuture.supplyAsync(()-> handler.runCompletion(doc, pos, offset));
         }
-        return CompletableFuture.allOf(futures).thenApply(ignored->{
-            // coalesce results.  This _really_ should, ideally, just be just pushed to clients...
-            Set<CompletionFragment> all = new LinkedHashSet<>();
-            for (Collection<CompletionFragment> result : results) {
-                all.addAll(result);
+        Set<CompletionItem.Builder> all = new LinkedHashSet<>();
+        for (CompletableFuture<Collection<CompletionItem.Builder>> future : futures) {
+            try {
+                all.addAll(future.get(5, TimeUnit.SECONDS));
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                break;
+            } catch (ExecutionException e) {
+                // TODO: log this;
```

Comment:
file issues for new TODOs, or link them in code - format is `// TODO (repo#issuenum) message`

Reply:
I just stuck in a logger, and trace() log it, instead of throwing an exception and breaking things over autocomplete misbehaving.